taichi/analysis/gather_statements.cpp

```diff
@@ -0,0 +1,38 @@
+#include "taichi/ir/ir.h"
+
+TLANG_NAMESPACE_BEGIN
+
+class StmtSearcher : public BasicStmtVisitor {
+ private:
+  std::function<bool(Stmt *)> test;
+  std::vector<Stmt *> results;
+
+ public:
+  using BasicStmtVisitor::visit;
+
+  StmtSearcher(std::function<bool(Stmt *)> test) : test(test) {
+    allow_undefined_visitor = true;
+    invoke_default_visitor = true;
+  }
+
+  void visit(Stmt *stmt) {
```

Comment:
Does it override `visit` methods of container statements in `BasicStmtVisitor`? If the answer is yes, it may be unable to gather statements recursively over container statements.

Reply:
I just wrote a toy example and found that `void visit(Stmt *stmt)` only affects types that can be converted to `Stmt *` and without corresponding non-virtual methods in ancestor classes. So it can gather statements recursively over container statements.

Reply:
> Does it override `visit` methods of container statements in `BasicStmtVisitor`? If the answer is yes, it may be unable to gather statements recursively over container statements.

It only overrides the generic `visit(Stmt *)` method. Container statements are treated specially in `BasicStmtVisitor`.